p2panda-rs/src/storage_provider/traits/storage_provider.rs

```diff
@@ -0,0 +1,484 @@
+                        .entry_decoded()
+                        .seq_num()
+                        .clone()
+                        .next()
+                        .unwrap(),
+                    *entry_latest.entry_decoded().log_id(),
+                ))
+            }
+            // No entry was given yet, we can assume this is the beginning of the log
+            None => Ok(Self::EntryArgsResponse::new(
+                None,
+                None,
+                SeqNum::default(),
+                log,
+            )),
+        }
+    }
+
+    /// Implementation of `panda_publishEntry` RPC method.
+    ///
+    /// Stores an author's Bamboo entry with operation payload in database after validating it.
+    async fn publish_entry(
+        &self,
+        params: &Self::PublishEntryRequest,
+    ) -> Result<Self::PublishEntryResponse, StorageProviderError> {
+        // Create an `EntryWithOperation` which also validates the encoded entry and operation.
+        let entry_with_operation = EntryWithOperation::new(
+            params.entry_encoded().to_owned(),
+            params.operation_encoded().to_owned(),
+        )?;
+
+        // Decode author, entry and operation. This conversion validates the operation hash
+        let author = params.entry_encoded().author();
+        let entry_encoded = params.entry_encoded();
+        let entry = decode_entry(params.entry_encoded(), Some(params.operation_encoded()))?;
+        let operation = Operation::from(params.operation_encoded());
+
+        // Every operation refers to a document we need to determine. A document is identified by the
+        // hash of its first `CREATE` operation, it is the root operation of every document graph
+        let document_id = if operation.is_create() {
+            // This is easy: We just use the entry hash directly to determine the document id
+            DocumentId::new(entry_encoded.hash().into())
+        } else {
+            // For any other operations which followed after creation we need to either walk the operation
+            // graph back to its `CREATE` operation or more easily look up the database since we keep track
+            // of all log ids and documents there.
+            //
+            // We can determine the used document hash by looking at what we know about the previous
+            // entry in this author's log.
+            //
+            // @TODO: This currently looks at the backlink, in the future we want to use
+            // "previousOperation", since in a multi-writer setting there might be no backlink for
+            // update operations! See: https://github.com/p2panda/aquadoggo/issues/49
+            let backlink_entry_hash = entry
+                .backlink_hash()
+                .ok_or(PublishEntryError::OperationWithoutBacklink)?;
+
+            self.get_document_by_entry(backlink_entry_hash)
+                .await?
+                .ok_or(PublishEntryError::DocumentMissing)?
```

Comment:
This error trips if the backlink is missing, not necessarily the actual document. With the move to checking previous operations this will change, but still relevant to keep in mind.

Reply:
Could be transformed to `PublishEntryError::BacklinkMissing` here.